File: Begleitmaterial/Taylor_Green_Wirbel.m
%% Visualisierung des Taylor-Green-Wirbels

function Taylor_Green_Wirbel(options)
arguments
    options.nu   double = 0   % Kinematische Viskosität
    options.CFL  double = 1   % CFL-Zahl für den Zeitschritt
    options.x_nr uint16 = 100 % Auflösung in x-Richtung
    options.y_nr uint16 = 20  % Auflösung in y-Richtung
end

% Gitter
x     = linspace(0,2*pi,options.x_nr);
y     = linspace(0,2*pi,options.y_nr);
dx    = x(2)-x(1);
dy    = y(2)-y(1);
[X,Y] = meshgrid(x,y);

% Lösung
U     = @(t) sin(X).*cos(Y).*exp(-2*options.nu*t);
V     = @(t) -cos(X).*sin(Y).*exp(-2*options.nu*t);
Psi   = @(t) sin(X).*sin(Y).*exp(-2*options.nu*t);

% Lagrange Partikel
subplot(1,2,1);
X_pos = X;
Y_pos = Y;
sctr  = scatter(X_pos',Y_pos','filled');
xlim([0 2*pi]);
ylim([0 2*pi]);
title('Lagrange Partikel');
xlabel('$x$','Interpreter','latex');
ylabel('$y$','Interpreter','latex');

% Stromfunktion
subplot(1,2,2);
cont = surfc(X,Y,Psi(0));
xlim([0 2*pi]);
ylim([0 2*pi]);
zlim([-1 1]);
title('Stromfunktion');
xlabel('$x$','Interpreter','latex');
ylabel('$y$','Interpreter','latex');
zlabel('$\Psi$','Interpreter','latex');

% Zeitschleife
t = 0;
while t <= 5

    % Aktuelle Strömungsgrößen
    U_now   = U(t);
    V_now   = V(t);
    Psi_now = Psi(t);
    
    % Plot
    for row = 1:options.y_nr
        sctr(row).XData = X_pos(row,:);
        sctr(row).YData = Y_pos(row,:);
    end
    cont(1).ZData = Psi_now;
    cont(2).ZData = Psi_now;
    sgtitle([ ...
        '$\nu=' num2str(options.nu) ...
        ',~t=' num2str(t,'%.2f') '$'],'Interpreter','latex');
    drawnow;
    
    % Nächster Zeitschritt
    t_stp = options.CFL ...
        / (max(abs(U_now),[],'all')/dx + max(abs(V_now),[],'all')/dy);
    t     = t + t_stp;
    X_pos = X_pos + t_stp*interp2(X,Y,U_now,X_pos,Y_pos);
    Y_pos = Y_pos + t_stp*interp2(X,Y,V_now,X_pos,Y_pos);

end
end

File: Begleitmaterial/Taylor_Green_Wirbel.py
import numpy as np
from scipy.interpolate import RectBivariateSpline
import matplotlib.pyplot as plt


""" Visualisierung des Taylor-Green-Wirbels """

def Taylor_Green_Wirbel(
    nu:    float = 0,   # Kinematische Viskosität
    t_end: float = 5,   # Endzeit
    t_nr:  int   = 100, # Anzahl der Zeitschritte
    x_nr:  int   = 100, # Auflösung in x-Richtung
    y_nr:  int   = 20   # Auflösung in y-Richtung
):

    # Gitter
    x    = np.linspace(0,2*np.pi,x_nr)
    y    = np.linspace(0,2*np.pi,y_nr)
    dx   = x[1]-x[0]
    dy   = y[1]-y[0]
    X, Y = np.meshgrid(x,y)

    # Lösung
    U    = lambda t: np.sin(X)*np.cos(Y)*np.exp(-2*nu*t)
    V    = lambda t: -np.cos(X)*np.sin(Y)*np.exp(-2*nu*t)
    Psi  = lambda t: np.sin(X)*np.sin(Y)*np.exp(-2*nu*t)
    
    plt.ion()
    fig = plt.figure(figsize=(6,4))

    # Lagrange Partikel
    ax    = fig.add_subplot(1,2,1)
    X_pos = np.copy(X)
    Y_pos = np.copy(Y)
    sctr  = [ax.scatter(X_pos[row],Y_pos[row],s=2) for row in range(y_nr)]
    ax.set_xlim(0,2*np.pi)
    ax.set_ylim(0,2*np.pi)
    ax.set_title('Lagrange Partikel')
    ax.set_xlabel(r'$x$')
    ax.set_ylabel(r'$y$')

    # Stromfunktion
    ax   = fig.add_subplot(1,2,2,projection='3d')
    surf = ax.plot_surface(X,Y,Psi(0),cmap=plt.cm.viridis)
    cont = ax.contour(X,Y,Psi(0),zdir='z',offset=-1,cmap=plt.cm.viridis)
    ax.set_xlim(0,2*np.pi)
    ax.set_ylim(0,2*np.pi)
    ax.set_zlim(-1,1)
    ax.set_title('Stromfunktion')
    ax.set_xlabel(r'$x$')
    ax.set_ylabel(r'$y$')
    ax.set_zlabel(r'$\Psi$')

    # Zeitschleife
    time  = np.linspace(0,t_end,t_nr)
    t_stp = time[1]-time[0]
    for t in time:

        # Aktuelle Strömungsgrößen
        U_now   = U(t)
        V_now   = V(t)
        Psi_now = Psi(t)
        
        # Plot
        for row in range(y_nr):
            sctr[row].set_offsets(np.stack((X_pos[row],Y_pos[row])).T)
        surf.set_zdata = Psi_now
        cont.set_zdata = Psi_now
        fig.suptitle(fr'$\nu={nu},~t={t:.2f}$')
        fig.tight_layout(pad=2)
        fig.canvas.draw()
        fig.canvas.flush_events()
        
        # Nächster Zeitschritt
        X_pos += t_stp * RectBivariateSpline(x,y,U_now.T)(X_pos,Y_pos,grid=False)
        Y_pos += t_stp * RectBivariateSpline(x,y,V_now.T)(X_pos,Y_pos,grid=False)
